lib/SILGen/SILGenApply.cpp

```diff
@@ -4549,12 +4585,34 @@ RValue SILGenFunction::emitApply(ResultPlanPtr &&resultPlan,
     emitForeignErrorCheck(loc, directResults, errorTemp, doesNotThrow,
                           *foreignError);
   }
-  
+
+  SmallVector<CleanupHandle, 8> cleanupHandles;
+  // For objc async calls, push cleanup to be used on throw paths in the result
+  // planner.
+  if (calleeTypeInfo.foreign.async) {
+    for (auto unmanagedCopy : unmanagedCopies) {
+      Cleanups.pushCleanup<FixLifetimeDestroyCleanup>(unmanagedCopy.getValue());
+      cleanupHandles.push_back(Cleanups.getTopCleanup());
+    }
+  }
+
   auto directResultsArray = makeArrayRef(directResults);
   RValue result =
     resultPlan->finish(*this, loc, substResultType, directResultsArray);
   assert(directResultsArray.empty() && "didn't claim all direct results");
 
+  // For objc async calls, generate cleanup on the resume path here and forward
+  // the previously pushed cleanups.
+  if (calleeTypeInfo.foreign.async) {
+    for (auto unmanagedCopy : unmanagedCopies) {
```

Comment:
Then here you can just emit and then forward the cleanup on the managed value. That is:

```
for (auto mv : unmanagedCopies) {
   auto value = mv.forward(*this);
   B.emitFixLifetime(...);
   B.emitDestroyOperation(...);
}
```

Reply:
Done